int32[] rate
int32[] direction